### DISCUSSION TITLE
Proposal: Empty String Coalescing Similar to Null Coalescing

### DISCUSSION THREAD

**Discussion by DashRader** [dotnet/csharplang#7280] [Language Ideas]
I would love to see an empty string coalescing operator to make comparisons easier.  This operator could be used in any similar way that ?? and ??= are used.

In c#, I have to do this:

```
myString = SomeExpressionThatMayReturnAnEmptyString();
myString = myString == "" ? " No Value" : myString;
  -OR-
myString = myString.IsNullOrBlank() ? "No Value" : myString;
  -OR-
myString = myString.IsNullOrEmpty() ? "No Value" : myString;
```

I would rather there be a simple operator instead such as $$ where $$ is the equivalent of ?? except for blank/empty. Think of it as empty string-coalescing such as $$ and $$=.

```
// If function returns "", then set myString = "No Value"
myString = SomeExpressionThatMayReturnAnEmptyString() $$ "No Value";
```
or even like this:

```
myString = SomeExpressionThatMayReturnAnEmptyString();
// If function returns "", then set myString = "No Value"
myString $$= "No Value";
```

**Comment by nuiva** [comment]
You could also do:
```c#
var myString = SomeExpressionThatMayReturnAnEmptyString() switch {
	""    => "No Value",
	var s => s
};
```
I think representing "no value" as empty string is code smell to begin with, so I don't like the proposal. Hopefully option types land soon.

**Reply by DashRader** [reply]
"No Value" is just used in my example, of course, it would be whatever you want it to be.  My point is to have an operator that does the coalesing for an empty string, just like ?? coalesces for null.  And I was thinking that using $$ would be a nice semantic operator for empty string checking.

**Comment by alrz** [comment]
You can use an extension method for that:
```cs
public static NullIfEmpty(this string str) => string.IsNullOrEmpty(str) ? null : str;
```
```cs
var x = str.NullIfEmpty() ?? "xyz";
```

**Reply by DashRader** [reply]
This is understood.  Point being I think that it would be very nice feature for it to be a native operation.

var x = str $$ "some default value if empty";

I mean, why does ?? exist?  Surely you could have written null check code like this:

```
public static NVL(this string str1, string str2) => str1 == null ? str2 : str1;
```

```
var x = str.NVL("use this string if null")
```

I could even write a similar one for empty:
```
public static EVL(this string str1, string str2) => str1 == "" ? str2 : str1;
```
```
var x = str.EVL("use this string if empty")
```

Maybe the $$ could also imply nulls as part of the empty check as well.

```
myString = ExpressionThatMayReturnAnEmptyOrNullString() $$ "Use this value if null or empty";
```

**Reply by DashRader** [reply]
BTW, when using ?? with a method, i.e. var x = Method() ?? "value", does the Method get executed only one time or does it get executed twice, once to check the value and once to actually retreive the value?

**Reply by HaloFour** [reply]
> BTW, when using ?? with a method, i.e. var x = Method() ?? "value", does the Method get executed only one time or does it get executed twice, once to check the value and once to actually retreive the value?

The method is only executed once.

**Reply by DashRader** [reply]
That's what I thought.  That's why I also like my suggestion.  No extra extensions to maintain.  No extra methods to run from a tight loop.  Let the compiled and faster code handle the comparison.

**Reply by CyrusNajmabadi** [reply]
I would expect no difference in speed between an extension and writing out the IL directly. :-)

**Comment by theunrepentantgeek** [comment]
> I mean, why does ?? exist? 

With **null** there's exactly one possible value to consider, and it's known to be a problem.

With your proposed `$$` operator there are many choices for what might be wanted on the left-hand side: 
* `""` exactly
* `""` or **null** - aka where `IsNullOrEmpty()` returns **true**
* Any string where `IsNullOrBlank()` returns **true**

You've allluded to all three of these alternatives yourself. There are other possibilities too. Some systems use deliberate sentinel values to indicate the field is empty - so there are cases where `"-"` or `"n/a"` or even `"null"` might be encountered.

At this point it's a matter of _policy_ what you want to collapse - and policy changes, from developer to developer, from project to project, and from company to company. 

Even if there was will from the C# team to implement this, there's literally no right answer for what the feature would do.

By writing your own extension method, you can have the policy you want:

``` csharp
// If function returns "", then set myString = "No Value"
myString = SomeExpressionThatMayReturnAnEmptyString().DefaultTo("No Value");
```

Not only does this have the benefit of being something you can have today (instead of waiting several years because all language changes are carefully considered before implementation), but it's arguably clearer code than some cryptic `$$` sigil that's totally ungoogleable.

**Comment by KennethHoff** [comment]
I think a more general solution to this would be being able to overwrite what counts as "null" for the sake of null-coalescing and not just for strings.

For example, it would be nice to be able to say:

```c#
private readonly record struct UserId(Guid UnderlyingValue)
{
	// Pseudo-code
	public static override bool CountsAsNull(UserId value)
	{
		return value.UnderlyingValue == Guid.Empty
	}
}

private static UserId GetId()
{
	// Returns a non-nullable UserId that can be equivalent to Guid.Empty
}

var userId = GetId() ?? UserId.System; // Either a valid non-Empty Guid, or UserId.System
```

This specific case would be better served by fixing `GetId()` to simply return null instead in the Empty-case, but there are scenarios where this would simplify things.. It would also confuse things though, so probably not worth it
